### ISSUE TITLE
Specifying the CRAN environmental variable doesn't seem to do anything?

### ISSUE THREAD

**Issue by andburch** [rocker-org/rocker-versioned2#591]
### Container image name

rocker/tidyverse:4.2.2

### Container image digest

rocker/tidyverse:sha256:85f108b64aaac0280d8c9ca85831ae0ce351f7afc16d23fb9859f23ef4b1cc10

### What operating system are you seeing the problem on?

Windows

### Bug description

My workplace has a very restrictive firewall/web-filtering program, so I can't access sites like https://packagemanager.posit.co. I DID manage to get access to the "default CRAN" site (https://cran.r-project.org), which is what I use to download packages. Normally, I would run some code to make sure I'm using the correct source: 

```
local({r <- getOption("repos")
r["CRAN"] <- "https://cran.r-project.org"
options(repos=r)
rm(r)
})
```
I thought that specifying the "CRAN" environmental variable in the `docker run` command would automatically do something like this, but it doesn't seem to work. I keep getting the usual `Warning in install.packages :
  unable to access index for repository https://packagemanager.posit.co/cran/__linux__/jammy/latest/src/contrib:
  cannot open URL 'https://packagemanager.posit.co/cran/__linux__/jammy/latest/src/contrib/PACKAGES'` errors when I try to install new packages. Am I understanding this incorrectly and should I just run the above code as a script every time I start a new container?

### How to reproduce this bug?

```sh
docker run -e PASSWORD=bababooey --env=PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin --env=R_VERSION=4.2.2 --env=R_HOME=/usr/local/lib/R --env=TZ=Etc/UTC --env=CRAN=https://cran.r-project.org --env=LANG=en_US.UTF-8 --env=S6_VERSION=v2.1.0.2 --env=RSTUDIO_VERSION=2022.12.0+353 --env=DEFAULT_USER=rstudio --env=PANDOC_VERSION=default --env=QUARTO_VERSION=default -p 8787:8787 --runtime=runc -d rocker/tidyverse:latest
```

**Comment by eddelbuettel** [comment]
The snippet

```r
local({r <- getOption("repos")
r["CRAN"] <- "https://cran.r-project.org"
options(repos=r)
rm(r)
})
```

respects an `options()` value of its `repos` variable for the named entry CRAN.  That is not the same as respecting an environment variable.  Also this whole segment is for a 'normal' R installation whereas ... the r-versioned stack underlying the tidyverse container uses a different source anyway...   So it's a bit of an impedance mismatch.

Am I understanding you correctly that you would like to use `install.packages()` as usual but pointing to a CRAN mirror instance?  You should always be able to do that explicitly too as in

```r
install.packages("digest", repos="http://cloud.r-project.org")
```

**Comment by andburch** [comment]
Yeah, this makes sense to me. But what does specifying the environmental variable of "CRAN" then actually change in the container?

**Comment by eddelbuettel** [comment]
I don't know.  Many many years of using (and contributing to) CRAN but I never drove it via an environment variable.

**Comment by andburch** [comment]
Okay, that's good enough for me. I'll figure it out via other means. Thank you!

**Comment by eddelbuettel** [comment]
Sounds good.  I'll close this then, feel free to reopen if something comes up.

**Comment by andburch** [comment]
Actually, one quick question:

In the r-ver dockerfile, there are lines that essentially do the same thing as what I was describing (but set the repos to a MRAN snapshot, which is even BETTER for my case):

```
...
&& if [ -z "$BUILD_DATE" ]; then MRAN=$CRAN; \
   else MRAN=https://mran.microsoft.com/snapshot/${BUILD_DATE}; fi \
   && echo MRAN=$MRAN >> /etc/environment \
  && echo "options(repos = c(CRAN='$MRAN'), download.file.method = 'libcurl')" >> /usr/local/lib/R/etc/Rprofile.site \
...
```

But even if I specify a BUILD_DATE environmental variable, when I check R the "repos" CRAN option is still set to "https://packagemanager.posit.co/cran/__linux__/jammy/latest" somehow. 

I don't think this is how it is supposed to work at all: where is the the posit.co URL even being introduced and why is it overriding the MRAN functionality?

**Comment by andburch** [comment]
[The only place I can find that URL in the repository](https://github.com/rocker-org/rocker-versioned2/search?q=packagemanager.posit.co%2Fcran%2F__linux__%2Fjammy%2Flatest&type=code) is in the r-ver dockerfile (where it's given as a default environmental variable) or in the relevant "stacks" file:

```
...
     "COPY_a_script": "scripts/install_R_source.sh /rocker_scripts/install_R_source.sh",
      "RUN_a_script": "/rocker_scripts/install_R_source.sh",
      "ENV_after_a_script": {
        "CRAN": "https://packagemanager.posit.co/cran/__linux__/jammy/latest",
        "LANG": "en_US.UTF-8"
      },
...
```

Should it be automatically setting the CRAN environmental variable to something that the user can't control? That seems "not good" to me...

**Comment by eitsupi** [comment]
Please check the reference page.
https://rocker-project.org/images/versioned/r-ver.html#how-to-use

**Comment by andburch** [comment]
Ah! So the BUILD_DATE variable is _not_ supposed to be used as the authoritative way to set up a MRAN snapshot? Just to be clear, it's _not_ there for me to use to manage packages etc?

**Comment by andburch** [comment]
> Also note that MRAN will not extend recording beyond Jan 2023, and be turned off July 2023.

Wow, I did _not_ realize this. Thank you so much for saving me future headaches!

**Comment by eitsupi** [comment]
Those environment variables are variables that are referenced when building Docker images and are not intended for user use.
`CRAN` is only used in the following script and written to the `Rprofile` file.

https://github.com/rocker-org/rocker-versioned2/blob/23b6961dcc187b7290b35ae4434180dfb1fa7f24/scripts/setup_R.sh#L38

https://github.com/rocker-org/rocker-versioned2/blob/23b6961dcc187b7290b35ae4434180dfb1fa7f24/dockerfiles/r-ver_devel.Dockerfile#L21

In other words, running this script after changing the `CRAN` environment variable will change the repository that R refers to, but there is no guarantee that the `CRAN` is a valid value because this script does not validate the URL.

If you want to use a CRAN mirror for any date, the `pak` package may be what you are looking for.
https://pak.r-lib.org/reference/repo_add.html

**Comment by andburch** [comment]
I'm sorry for sucking away a bit of your time, but I am very thankful for your help! I'm relatively new to Docker/Rocker and _very_ new to orchestrating package management across multiple machines/containers. You helped resolve an impending crisis that I didn't know I had! Thank you for everything.

**Comment by eitsupi** [comment]
Never mind about the question.

RSPM takes snapshots of CRAN almost daily, so basically it should be able to be used instead of MRAN. The main difference is that no binary packages are provided for macOS, while binary packages are provided for some Linux distributions.

An easy way to check if a CRAN snapshot exists for a given date is to use the pak package.
This repository also uses the function from the pak package to verify CRAN URLs.

https://github.com/rocker-org/rocker-versioned2/blob/23b6961dcc187b7290b35ae4434180dfb1fa7f24/build/make-stacks.R#L70-L72

Of course you can also select snapshots in the Package Manager's web UI.
https://packagemanager.rstudio.com/client/#/repos/2/overview
